### PR TITLE
SASL GSSAPI/Kerberos support

### PR DESCRIPTION
**Pull request by jsurloppe**
Hi,

I've take the opportunity that's you're currently working on SASL_PLAIN to add GSSAPI to your branch.

I've tested my branch on confluent cp-docker-images/examples/kafka-cluster-sasl (PLAINTEXT) and our production cluster (SSL), still doing some tests but i'm not encountering any problem for now.

### CODE REVIEW

File: aiokafka/conn.py [hunk 1]
```diff
@@ -18,10 +19,18 @@
 import aiokafka.errors as Errors
 from aiokafka.util import ensure_future, create_future, PY_341, PY_36
 
+from kafka.protocol.types import Int8
```
Comment (tvoinarovskyi) [inline aiokafka/conn.py]: Please just use `struct` package directly

File: aiokafka/conn.py [hunk 2]
```diff
@@ -262,8 +289,39 @@ def _do_sasl_handshake(self):
...
+        server_token = None
+        while not client_ctx.complete:
+            client_token = client_ctx.step(server_token)
+            if client_token:
+                server_token = yield client_token, True
+            else:
+                server_token = yield b'', True
+        msg = client_ctx.unwrap(server_token).message
+
+        iomsg = io.BytesIO(msg[0:1])
```
Comment (tvoinarovskyi) [inline aiokafka/conn.py]: Int8 is just 1 byte, right? msg[0] will give you the value already. This is full Python3, so no need for this wrapping.

**Comment by jsurloppe** [comment]
Totally agree that we should have a docker test env with all the stack for testing againt next versions.

For now the thing the only thing that annoy me is that gssapi client_ctx.step() trigger indirect blocking IO (krb libs exchanging with KDC), it's only used for ticket exchange so i'm not sure if it's worth an executor...

**Comment by tvoinarovskyi** [comment]
It is worth an executor. It's a network call, right? So if any kind of network issue happens it means we will end up with a stop in our application. That is no good.

File: aiokafka/conn.py [hunk 3]
```diff
@@ -216,7 +233,11 @@ def _do_version_lookup(self):
 
     @asyncio.coroutine
     def _do_sasl_handshake(self):
-        req_klass = self._version_info.pick_best(SaslHandShakeRequest)
+        if self._sasl_mechanism == "GSSAPI":
+            # v1 is unsupported, trigger a infinite auth loop
```
Comment (tvoinarovskyi) [inline aiokafka/conn.py]: I feel like this is not true. The v1 is supported, we just need to figure out how it's done. It works with Java client )

**Comment by tvoinarovskyi** [comment]
@jsurloppe As I see this is based on kafka-python implementation. I know the author spent many hours there setting that up, so I feel like we can go with this implementation, it's still better than none. So we can delegate automatic testing of this feature, as for executor and v1 support I can totally merge without them if you feel like those would be hard to implement.

File: aiokafka/conn.py (hunk 3 above)
Comment (jsurloppe) [reply aiokafka/conn.py]: dunno, both kafka-python and rdkafka use v0

**Comment by jsurloppe** [comment]
I tried different approach but the executor is pretty tricky, futures don't like to be passed between loops and generated based coroutine won't allow async generator.

We might be able to have something cleaner if we move to new syntax and async gen 3.6+

As for v1, not sure to find the time to investigate at the moment, tbh.

**Comment by tvoinarovskyi** [comment]
@jsurloppe Cool, thanks. Will merge this and do a followup after if any fixes needed. I have no idea why travis did not build for a fork thou...

File: aiokafka/conn.py (hunk 3 above)
Comment (tvoinarovskyi) [reply aiokafka/conn.py]: Lol, version v1 did not work for such a dumb reason. Basically a few of my buggy lines of code where req_klass is used in 2 places and next request just always ended up with API_VERSION=0 =). It works for both versions, kafka-python and rdkafka just did not pick this up yet.

**Pull request merged**